src/SelfUpdateManager.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace SelfUpdate;
+
+use Composer\Semver\Comparator;
+use Composer\Semver\VersionParser;
+use Composer\Semver\Semver;
+use GuzzleHttp\HandlerStack;
+use Kevinrob\GuzzleCache\CacheMiddleware;
+use Kevinrob\GuzzleCache\Storage\Psr6CacheStorage;
+use Kevinrob\GuzzleCache\Strategy\PublicCacheStrategy;
+use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+use UnexpectedValueException;
+
+/**
+ * Business logic for the self-update command.
+ *
+ * @author Alexander Menk <alex.menk@gmail.com>
+ */
+class SelfUpdateManager
+{
+    protected array $options;
+    private ?array $latestRelease = null;
+
+    public function __construct(protected string $gitHubRepository, protected string $currentVersion, protected string $applicationName, array $options){
+        // Options need to be passed in the constructor since we cache the latestRelease which may vary based on options.
+        $this->options = array_merge([
+            'preview' => false,
+            'compatible' => false,
+            'version_constraint' => null,
+        ], $options);
+    }
+
+    /**
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function isUpToDate(): bool {
+        $latestRelease = $this->getLatestReleaseFromGithub();
+        return NULL === $latestRelease || Comparator::greaterThanOrEqualTo($this->currentVersion, $latestRelease['version']);
+    }
+
+    /**
+     * Get the latest release version and download URL according to given
+     * constraints.
+     *
+     * @return string[]|null
+     *    "version" and "download_url" elements if the latest release is
+     *     available, otherwise - NULL.
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function getLatestReleaseFromGithub(): ?array
+    {
+        if (null !== $this->latestRelease) {
```

Comment:
I can stop caching this at the class level now that we cache the network request. Maybe the whole class could be made static then.